### PR TITLE
Migrate RAG pipeline to async processing.

### PR DESCRIPTION
**Pull request by austintlee**
### Description
Use the async version of the search pipeline process to avoid blocking (remote) calls.

Original bug - https://github.com/opensearch-project/OpenSearch/issues/10248. 
 
### Issues Resolved
- #2334 
 
### Check List
- [x ] New functionality includes testing.
  - [x ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by austintlee** [comment]
cc: @reta @msfroh

**Comment by austintlee** [comment]
@msfroh Can you take a look at my code changes?  I have a really basic question about exception handling in `processResponseAsync` - if I want to throw things like IllegalArgumentException/InvalidInputException, do I just throw it and let it propagate out of the method or is there some contract I need to follow in this async world?

**Review by Zhangxunmt — COMMENTED** [review]
A lot of "onFailure(Exception e)" implementations in the ActionListeners do not include log.error(). Should we add more logs for errors? This was brought up in the earlier security review too.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:135]: Changed to debug.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:147]: Removed.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:153]: Yes

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:171]: Yes

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:197]: Changed to debug.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:234]: Changed to debug.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/llm/DefaultLlmImpl.java:92]: Done.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:122]: Changed to debug.

**Review by ylwu-amzn — APPROVED** [review]
Thanks for the quick fix

**Review by Zhangxunmt — APPROVED** [review]

**Comment by msfroh** [comment]
> @msfroh Can you take a look at my code changes? I have a really basic question about exception handling in `processResponseAsync` - if I want to throw things like IllegalArgumentException/InvalidInputException, do I just throw it and let it propagate out of the method or is there some contract I need to follow in this async world?

Shoot -- I didn't see your question before this got merged. (I spent most of yesterday traveling to a conference.)

The general contract in the async world is that every possible code path needs to notify the listener *exactly once* or else it will wait indefinitely.
* If you have a response, you must give it to the listener.
* If there's a failure, you must notify the listener (probably via `onFailure`).
* If there's no response, you must notify the listener that there's no response.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:145]: Yes, I have a test for this, but it does not go through the REST layer.  I may need an IT test.

File: search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java
Comment (austintlee) [reply search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/GenerativeQAResponseProcessor.java:308]: I'll check and also test.

**Pull request merged**
